### PR TITLE
Jatin improve report loading time

### PR DESCRIPTION
**Pull request by JatinAgrawal94**
# Description
Screenshot 2024-09-25 at 11 02 20 PM

## Related PRS (if any):
To test this backend PR you need to checkout the [#1132](https://github.com/OneCommunityGlobal/HGNRest/pull/1132) backend PR.
…

## Main changes explained:
- Update file src/components/Reports/TotalTeamReport.jsx for improving load time.

## How to test:
1. check into current branch
2. do `npm install` and `npm run start:local` to run this PR locally
3. Clear site data/cache
4. log as owner user
5. go to dashboard→ Reports→ Reports→Show Total Team Report
6. The original graph in development version takes 30 seconds to load. While reviewing this PR please keep a watch on time it takes to load that graph.

## Screenshots or videos of changes:

https://github.com/user-attachments/assets/c1454060-8f0e-41df-a8ae-041d133bc1e3

## Note:
Include the information the reviewers need to know.

### CODE REVIEW

**Review by slidracoon72 — APPROVED** [review]
#2780 
@JatinAgrawal94 
I have tested this PR, and the report loading time has noticeably improved. The initial load took approximately 3-4 seconds, while subsequent loads were significantly faster in comparison.

Video Demonstration:
https://www.loom.com/share/fb9d21bc24c448018d49b2cf1bc89327?sid=a8e55ca0-8ec0-42eb-8596-6501f2a96a41

File: src/components/Reports/Reports.jsx [hunk 1]
```diff
@@ -309,6 +309,7 @@ class ReportsPage extends Component {
   }
 
   showTotalProject() {
+    console.log("I am I being called?")
```
Comment (lng2020) [inline src/components/Reports/Reports.jsx:312]: Forget to delete?

File: src/components/Reports/TotalReport/TotalProjectReport.jsx [hunk 2]
```diff
@@ -11,7 +11,7 @@ import { set } from 'lodash';
 
 function TotalProjectReport(props) {
   const { startDate, endDate, userProfiles, projects, darkMode } = props;
-
+  console.log("I am inside")
```
Comment (lng2020) [inline src/components/Reports/TotalReport/TotalProjectReport.jsx:14]: same

File: src/components/Reports/TotalReport/TotalProjectReport.jsx [hunk 3]
```diff
@@ -192,6 +192,7 @@ function TotalProjectReport(props) {
   );
 
   const totalProjectInfo = totalProject => {
+    console.log("TotalProjectinfo started");
```
Comment (lng2020) [inline src/components/Reports/TotalReport/TotalProjectReport.jsx:195]: same

File: src/components/Reports/TotalReport/TotalProjectReport.jsx [hunk 4]
```diff
@@ -248,7 +249,7 @@ function TotalProjectReport(props) {
       ) : (
         <div>
           <div>{totalProjectInfo(allProject)}</div>
-          <div>{showTotalProjectTable ? totalProjectTable(allProject) : null}</div>
+          {/* <div>{showTotalProjectTable ? totalProjectTable(allProject) : null}</div> */}
```
Comment (lng2020) [inline src/components/Reports/TotalReport/TotalProjectReport.jsx:252]: same

File: src/components/Reports/TotalReport/TotalTeamReport.jsx [hunk 5]
```diff
@@ -452,10 +487,11 @@ function TotalTeamReport(props) {
       ) : (
         <div>
           <div>{totalTeamInfo(allTeams)}</div>
-          <div>{showTotalTeamTable ? totalTeamTable(allTeams, userNameList) : null}</div>
+          {/* <div>{showTotalTeamTable ? totalTeamTable(allTeams, userNameList) : null}</div> */}
```
Comment (lng2020) [inline src/components/Reports/TotalReport/TotalTeamReport.jsx:490]: same

**Review by lng2020 — COMMENTED** [review]
And please resolve the conflict. Otherwise the icon is missing. I think should be fine in the latest development branch.
Screenshot_20241011_220841

**Review by shivansh-nathani — COMMENTED** [review]
This are working as expected.
Few changes that can be made before merging

1. Remove logs for testing and debugging
2. 'i' icon is missing next to 'Show missing team report' (tried to trace where we lost it but couldn't find it using prev two blame's, not sure if the removal was intentional, we can add it back if someone removed that by-mistake)
Screenshot 2024-10-11 at 8 39 08 PM

**Review by Jingyii800 — APPROVED** [review]
I have tested this function, the subtask bar keeps dark mode after opening it. Great job!

https://github.com/user-attachments/assets/68db3fb6-e7c4-44c8-8108-385e9f2a643f

**Review by lucy-xfl — DISMISSED** [review]
function works well in my side.

File: src/components/Reports/Reports.jsx [hunk 6]
```diff
@@ -598,26 +598,11 @@ class ReportsPage extends Component {
...
-                        fontSize={15}
-                        isPermissionPage
-                        darkMode={darkMode}
-                      />
-                    </div>
-                  </div>
-                </div>
+              </div>
+              </div>
                 </div>
```
Comment (lng2020) [inline src/components/Reports/Reports.jsx:606]: Is the deletion right?

Comment (JatinAgrawal94) [reply src/components/Reports/Reports.jsx:606]: I have tested the PR with some changes requested above. Please try and review again.

Comment (lng2020) [reply src/components/Reports/Reports.jsx:606]: The icon is missing. It's likely due to the accidental deletion of the code snippet while resolving merge conflicts. It's around line 600 in Reports.jsx as my review comment suggests.
Screenshot_20241012_031507

Comment (JatinAgrawal94) [reply src/components/Reports/Reports.jsx:616]: Thanks for pointing that out. I have added the Icon. Please let me know if you find something missing.

**Review by ishangoel17 — COMMENTED** [review]
Verified the PR and the changes are working as expected. Below are the screenshots for reference:
Capture2780

**Review by aaryaneil — APPROVED** [review]
Tested the PR and the screen loaded in under 20 seconds, good work.

https://github.com/user-attachments/assets/cf3b0077-6846-454b-923f-4382c0c4e865

**Comment by pallavithorat** [comment]
I have reviewed the PR, and the changes are functioning as expected. 
PR#2780

**Review by javid679 — APPROVED** [review]
Hi I have checked this PR and everything is working fine, but a slight change is appreciated in the UI, which might not related to you, the numbers showing on the bar are really close to the bar graph, and they need to be moved up a little bit. Nevertheless,  the loading time has been reduced significantly, good job. Below here is the screenshot for reference.

image

Working Screenshot:
Screenshot (62)

**Review by bhavya17prakash — APPROVED** [review]
Verified load time is improved (20 secs)

https://github.com/user-attachments/assets/a71a8248-f8ff-41d2-92f9-d82ae1b84c32

**Comment by geetamatkar** [comment]
Hi @JatinAgrawal94 ,
I tested this PR and it is working fine but it took more than 30 seconds to load graph.

Screenshot 2024-10-19 at 4 33 00 PM
Screenshot 2024-10-19 at 4 35 51 PM

**Review by hqzhu0913 — COMMENTED** [review]
I review the backend PR and frontend PR. I just have a question that it seems all the data processing is in the front end. May this the reason be the reason it's slow when loading the charts? Would the query to mongoose database use some aggregate functions and provide simpler data to the frontend.

**Comment by JatinAgrawal94** [comment]
> I review the backend PR and frontend PR. I just have a question that it seems all the data processing is in the front end. May this the reason be the reason it's slow when loading the charts? Would the query to mongoose database use some aggregate functions and provide simpler data to the frontend.

I checked that. Backend is the one taking more time to prepare data. And I did use some agregate functions to improve speed.

**Review by CodewithKoushica — CHANGES_REQUESTED** [review]
The changes have improved the load time, but it still takes around 1 minute, which is still quite long.

Suggestions for further improvement:

- Optimize the /users API calls to reduce redundancy and improve efficiency.
- Consider implementing lazy loading or caching to enhance performance.
- Add a more informative loading indicator to improve user experience during long load times.

Screenshot 2024-10-22 at 11 20 21 AM

**Review by samarth9201 — COMMENTED** [review]
Tested the PR, it still take about 20-25 seconds to load the graph for me. Not sure what the issue could be but needs further optimization.

Screenshot (106)
Screenshot (107)

**Review by slidracoon72 — DISMISSED** [review]
#2780 
@JatinAgrawal94 
I reviewed the PR again, and this time it took 1 minute and 20 seconds for the Team Report to load. Further optimization is necessary to improve the loading speed.

Time:
IMG_17580E437CE8-1

Screenshot 2024-10-24 at 2 50 43 PM

**Review by CodewithKoushica — CHANGES_REQUESTED** [review]
The recent updates show great progress in optimizing load time. Here’s a summary of what I observed:

**Front-end:**
- User calls have been reduced, and data caching has been successfully implemented, which greatly improves efficiency.
- However, the received data isn't always populating correctly on the front end. Although the data is retrieved, it doesn’t always render as expected.

**Back-end:**
- Occasionally, `/reports` calls fail and don’t return any data. This intermittent issue could impact the reliability of the reports display.

Thank you for addressing the initial feedback; these adjustments have definitely made improvements. Fixing the data rendering and stabilizing the backend API would make this feature more robust.

Screenshot 2024-10-25 at 11 15 52 AM
Screenshot 2024-10-25 at 11 16 22 AM

https://github.com/user-attachments/assets/5442163b-9582-40e2-a276-358be035262d

**Review by CodewithKoushica — COMMENTED** [review]
There is a good improvement with the timings but the report chart isn't generating.
Please check for that.
Screenshot 2024-10-30 at 9 51 03 AM

**Comment by JatinAgrawal94** [comment]
> There is a good improvement with the timings but the report chart isn't generating. Please check for that. Screenshot 2024-10-30 at 9 51 03 AM

Please try to clone the repo again, and run the app. It works fine for me on all the browsers.

**Review by hqzhu0913 — APPROVED** [review]
Great job! The load time is less than 5 seconds in my test. 
Screenshot 2024-10-31 at 2 17 30 PM

**Review by Rui-Liu-github — APPROVED** [review]
I reviewed PR #2780, and tested  the loading time of the the "show total team report" button. The loading time has improved and is now faster.
image

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**